stdlib/src/collections/string.mojo

```diff
@@ -1324,6 +1336,44 @@ struct String(
 
         return result
 
+    fn fast_join[
+        T: SizedSpanableCollectionElement, //,
+    ](self, elems: List[T, *_]) -> String:
+        """Joins string elements using the current string as a delimiter.
+
+        Parameters:
+            T: The types of the elements.
+
+        Args:
+            elems: The input values.
+
+        Returns:
+            The joined string.
+        """
+        var n_elems = len(elems)
+        if n_elems == 0:
+            return String("")
+        var len_self = self.byte_length()
+        var len_elems = 0
+        # Calculate the total size of the elements to join beforehand
+        # to prevent alloc syscalls as we know the buffer size.
+        # This can hugely improve the performance on large lists
+        for e in elems:
+            len_elems += len(e[])
+        var capacity = len_self * (n_elems - 1) + len_elems
+        var buf = Self._buffer_type(capacity=capacity)
+        var i = 0
+        var is_first = True
+        while i < n_elems:
+            if is_first:
+                is_first = False
+            else:
+                buf.extend(self.as_bytes_span())
+            buf.extend(elems[i].as_bytes_span())
+            i += 1
+        buf.append(0)
```

Comment:
It's actually x3 times faster. Fixed here: https://github.com/msaelices/mojo/commit/7f1fd84f69cedcecef61d34645a02fa304917710

I've kept some of the code as:
1. The branching overhead is minimal and makes the code more readable. Anyways, there will be a branch inside the first `memcpy` call, checking if `count` is zero.
2. The `offset += len_self` line cannot be executed on each iteration.
3. We still need the `i` variable to get the `elems[i]` obj.

Reply:
> It's actually x3 times faster. Fixed here: https://github.com/msaelices/mojo/commit/7f1fd84f69cedcecef61d34645a02fa304917710

wow didn't expect that much of an impact :fire: . How close are we to the same code performance in Python?

> I've kept some of the code as:
> 1. The branching overhead is minimal and makes the code more readable. Anyways, there will be a branch inside the first memcpy call, checking if count is zero.

I did it because you'll pay for the branching inside `memcpy` wether you branch with `is_first` or not :man_shrugging: but yeah it's probably negligible

> 2. The offset += len_self line cannot be executed on each iteration.

Could also just multiply by not_first

> 3. We still need the i variable to get the elems[i] obj.

Yeah that one slipped my mind

---
I'm not seing the part `buf.unsafe_set(capacity, 0)` in the code that I think will matter because append also has some overhead

---

> The branching overhead is minimal and makes the code more readable

PS: I like that take generally speaking, but I think this function in particular will be used quite a lot and performance should be as optimal as possible.